### ISSUE TITLE
if..else..if block generates empty geometry

### ISSUE THREAD

**Issue by DarwinSurvivor** [openscad/openscad#4529]
The behaviour between "if" and "if else if" is inconsistent in whether it generates nothing or empty geometry.

The following generates nothing
```
if(false)
  cube();
```

The following generates empty geometry (object of zero volume)
```
if(false)
  cube();
else if (false)
  cube();
```

This become very evident when wrapped in an intersection with another object.

The following renders a cylinder since it's an intersection of 1 object (the cylinder)
```
intersection() {
  cylinder();
  if (false)
    cube();
}
```

The following renders nothing because it's an intersection of 2 objects (the cylinder and the zero-volume object generated by the if..else..if block)
```
intersection() {
  cylinder();
  if (false)
    cube();
  else if (false)
    cube();
}
```

I believe an if..else..if block with all false conditions should create nothing so as to be consistent with a false if statement.

**Comment by jordanbrown0** [comment]
An `if` that evaluates a block generates geometry, a `group` node containing the results of evaluating the block.

```
if (false) { }
```
does not evaluate a block and so generates nothing.
```
if (true) { }
```
evaluates the block and so generates a group node, and the block generates nothing.
```
if (false) { }
else { }
```
evaluates the `else` block and so generates a group node, and the block generates nothing.

```
if (false) { }
else {
    if (false) { }
}
```
evaluates the `else` block, and so generates a `group` node, and then the second `if` generates nothing.

---

Net, it's consistent:  if the `if` evaluates a block, it generates geometry.

Is it ideal?  Maybe, maybe not, but it's been that way for a long time so we're stuck with it.

Note that if you want it to consistently generate geometry, you can always include an empty `else` block.

One case where this causes a problem is an else-if chain
```
if (c1) { ... }
else if (c2) { ... }
else if (c3) { ... }
```
where you might think that you would get nothing if none of the conditions are true, but since it's really parsed like so:
```
if (c1) { ... }
else {
    if (c2) { ... }
    else {
        if (c3) { ... }
    }
}
```
the `else` blocks generate `group` nodes.

We could introduce an explicit `elseif`:
```
if (c1) { ... }
elseif (c2) { ... }
elseif (c3) { ... }
```
which would generate a `group` node only if it evaluates a block.

**Comment by jordanbrown0** [comment]
Workaround:
```
if (c1) { ... }
if (!c1 && c2) { ... }
if (!c1 && !c2 && c3) { ... }
```

**Comment by jordanbrown0** [comment]
We could probably also parse `else if` specially, but changing that behavior would at least in theory be incompatible.

**Comment by jordanbrown0** [comment]
I would say "working as designed and nailed down by legacy", except for the if-elseif chain behavior, which "should" generate no nodes if none of the tests succeed, but instead generates an empty node.

**Comment by kintel** [comment]
Yeah, the issue here isn't really the empty geometry, but the surprising intersection behavior.

**Comment by jordanbrown0** [comment]
Ah, but which intersection behavior do you find surprising?

```
intersection() {
   cube();
    if (false) sphere();
}
```
the fact that the second child isn't empty geometry, and so the result isn't empty, or
```
intersection() {
    cube();
    if (false) sphere();
    else {}
}
```
the fact that the second child *is* empty geometry and so the result *is* empty?

I find the first more surprising; I prefer the notion that every module invocation (including "if") always generates exactly one node, that might happen to be empty.

**Comment by kintel** [comment]
Knowing how OpenSCAD works internally, the first is more surprising, but I'd think most users tend to expect both cases to yield a no-op, not an empty object.

**Comment by jordanbrown0** [comment]
I wrote up an argument for why if(false) should generate a node containing nothing, and so when intersected with something else would yield nothing.

But really there was no practical value to that argument, because changing the behavior is precluded by legacy.

I think the only thing that we might change here is to make it so that
```
if (cond1) op1();
else if (cond2) op2();
```
yields no nodes (like a single failing "if") if neither condition is true.
